File: src/lex/lexer.cpp
#include "lexer.h"
#include "../error/log.h"
#include <cstddef>

namespace lygos {
    std::ostream &operator<<(std::ostream &os, const Token &token) {
        os << STRINGIFY(token.type) << ": " << token.value;
        return os;
    }

    Lexer::Lexer(const char *program): src(program), curr(program[0]), index(0), line(1), line_index(0) {}

    std::vector<Token> Lexer::GetTokens() {
        Token token = {"", TokenType::Arrow, line};
        std::vector<Token> tokens;
        while ((token = NextToken()).type != TokenType::Eof) {
            tokens.push_back(token);
        }
        return tokens;
    }

    Token Lexer::AdvanceToken(Token token) {
        Advance();
        return token;
    }

    Token Lexer::LexNumber() {
        std::string value = {0};
        size_t start = line_index;
        while (std::isdigit(curr) || curr == '.') {
            value.push_back(curr);
            Advance();
        }
        value.erase(0, 1);
        if(value.find('.') != std::string::npos)
            return {value, TokenType::Float, line};

        return {value, TokenType::Integer, {line, start}};
    }

    Token Lexer::LexString() {
        std::string value;
        size_t start = line_index;
        while (curr != '\"') {
            value.push_back(curr);
            Advance();
        }
        Advance();
        return {std::move(value), TokenType::String, {line, start}};
    }

    Token Lexer::LexChar() {
        std::string value;
        size_t start = line_index;
        while (curr != '\'') {
            value.push_back(curr);
            Advance();
        }
        Advance();
        return {std::move(value), TokenType::Char, {line, start}};
    }

    Token Lexer::LexId() {
        std::string value = {0};
        size_t start = line_index;
        while (std::isalpha(curr) || std::isdigit(curr) ||curr == '_'){
            value.push_back(curr);
            Advance();
        }
        value.erase(0, 1);
        if(KeyWords.contains(value)) {
            return {value, KeyWords.at(value), {line, start}};
        }

        return {value, TokenType::Id, {line, start}};
    }

    Token Lexer::NextToken() {
        while (curr != '\0') {
            while (std::isspace(curr)) {
                if(curr == '\n') { line++; line_index = 0; }
                Advance();
            }

            switch (curr) {
                case '.': return AdvanceToken({".", TokenType::Dot, {line, line_index}});
                case ',': return AdvanceToken({",", TokenType::Comma, {line, line_index}});
                case ';': return AdvanceToken({";", TokenType::Semi, {line, line_index}});
                case ':': {
                    switch (src[index + 1]) {
                        case ':': Advance(); return AdvanceToken({"::", TokenType::OpScope, {line, line_index}});
                        default: return AdvanceToken({":", TokenType::Colon, {line, line_index}});
                    }
                }
                case '+': return AdvanceToken({"+", TokenType::OpPlus, {line, line_index}});
                case '-': {
                    switch (src[index + 1]) {
                        case '>': Advance(); return AdvanceToken({"->", TokenType::Arrow, {line, line_index}});;
                        default: return AdvanceToken({"-", TokenType::OpMinus, {line, line_index}});
                    }
                }
                case '*': return AdvanceToken({"*", TokenType::OpMul, {line, line_index}});
                case '/': {
                    switch (src[index + 1]) {
                        case '/': {
                            while(curr != '\n') {
                                Advance();
                            }
                        } break;
                        default: return AdvanceToken({"/", TokenType::OpDiv, {line, line_index}});
                    }
                }break;
                case '%': return AdvanceToken({"%", TokenType::OpMod, {line, line_index}});
                case '[': return AdvanceToken({"[", TokenType::BraceLeft, {line, line_index}});
                case ']': return AdvanceToken({"]", TokenType::BraceRight, {line, line_index}});
                case '{': return AdvanceToken({"{", TokenType::CurlyLeft, {line, line_index}});
                case '}': return AdvanceToken({"}", TokenType::CurlyRight, {line, line_index}});
                case '(': return AdvanceToken({"(", TokenType::ParanLeft, {line, line_index}});
                case ')': return AdvanceToken({")", TokenType::ParanRight, {line, line_index}});
                case '#': return AdvanceToken({"#", TokenType::Hash, {line, line_index}});
                case '$': return AdvanceToken({"$", TokenType::Dollar, {line, line_index}});
                case '<': {
                    switch (src[index + 1]) {
                        case '=': Advance(); return AdvanceToken({"<=", TokenType::OpEqEq, {line, line_index}});
                        default: return AdvanceToken({"<", TokenType::AngleLeft, {line, line_index}});
                    }
                }
                case '>': {
                    switch (src[index + 1]) {
                        case '=': Advance(); return AdvanceToken({">=", TokenType::OpEqEq, {line, line_index}});
                        default: return AdvanceToken({">", TokenType::AngleRight, {line, line_index}});
                    }
                }
                case '&': {
                    switch (src[index + 1]) {
                        case '&': Advance(); return AdvanceToken({"&&", TokenType::OpAnd, {line, line_index}});
                        default: return AdvanceToken({"&", TokenType::Ampercent, {line, line_index}});
                    }
                }
                case '|': {
                    switch (src[index + 1]) {
                        case '|': Advance(); return AdvanceToken({"||", TokenType::OpOr, {line, line_index}});
                        default: return AdvanceToken({"|", TokenType::Pipe, {line, line_index}});
                    }
                }
                case '!': {
                    switch (src[index + 1]) {
                        case '=': Advance(); return AdvanceToken({"!=", TokenType::OpNeEq, {line, line_index}});
                        default: return AdvanceToken({"!", TokenType::Bang, {line, line_index}});
                    }
                }
                case '=': {
                    switch (src[index + 1]) {
                        case '=': Advance(); return AdvanceToken({"==", TokenType::OpEqEq, {line, line_index}});
                        default: return AdvanceToken({"=", TokenType::Equals, {line, line_index}});
                    }
                }                case '\"': Advance(); return LexString();
                case '\'': Advance(); return LexChar();
                case '\0': return {"", TokenType::Eof, {line, line_index}};
                default:
                    if(std::isdigit(curr))
                        return LexNumber();
                    if(std::isalpha(curr))
                        return LexId();

                Log::Logger::Abort(fmt::format("found unknown char `{}`", (int)curr));
            }
        }
        return {"", TokenType::Eof, 0};
    }
}

File: src/lexer/lexer.rs
pub mod lexer {
    use super::super::super::types::{KeyWords, TokenType};
    use std::path::PathBuf;

    #[derive(Debug, Clone)]
    pub struct Loc {
        pub file: PathBuf,
        pub start: usize,
        pub end: usize,
    }

    impl Loc {
        pub fn new(file: PathBuf, start: usize, end: usize) -> Self {
            return Self {
                file,
                start,
                end,
            };
        }
    }

    impl std::ops::Add for Loc {
        type Output = Loc;
        fn add(self, rhs: Self) -> Self::Output {
            Self::new(self.file, self.start, rhs.end)
        }
    }

    #[derive(Debug, Clone)]
    pub struct Tagged<T> {
        inner: T,
        loc: Loc,
    }

    impl<T> Tagged<T> {
        pub fn new(loc: Loc, value: T) -> Tagged<T> {
            return Self{ loc, inner: value };
        }

        pub fn inner(&self) -> &T {
            return &self.inner;
        }

        pub fn inner_mut(&mut self) -> &mut T {
            return &mut self.inner;
        }

        pub fn loc(&self) -> &Loc {
            return &self.loc;
        }
    }

    impl From<Token> for Tagged<String> {
        fn from(value: Token) -> Self {
            Self{ inner: value.value, loc: value.loc }
        }
    }

    #[derive(Debug, Clone)]
    pub struct Token {
        pub value: String,
        pub typ: TokenType,
        pub loc: Loc,
    }

    impl Token {
        pub fn new(value: &str, typ: TokenType, loc: Loc) -> Self {
            return Self {
                value: String::from(value),
                typ,
                loc,
            };
        }
    }

    pub struct Lexer {
        pub file: PathBuf,
        src: Vec<char>,
        curr: char,
        index: usize,
        line_index: usize,
    }

    impl Lexer {
        pub fn from(logic: &str, file: &PathBuf) -> Self {
            let mut parser = Self {
                file: file.clone(),
                src: logic.chars().collect(),
                curr: ' ',
                index: 0,
                line_index: 0,
            };
            parser.src.push('\0');
            parser.curr = parser.src[0];
            return parser;
        }

        pub fn get_tokens(&mut self) -> Vec<Token> {
            let mut token = self.next_token();
            let mut tokens = vec![];
            loop {
                tokens.push(token);
                token = self.next_token();
                if token.typ == TokenType::Eof {
                    break;
                }
            }
            return tokens;
        }

        fn advance(&mut self) {
            self.index += 1;
            self.curr = self.src[self.index];
            self.line_index += 1;
        }

        fn number(&mut self) -> Token {
            let mut value: String = String::new();
            let start = self.line_index;
            while self.curr.is_digit(10) || self.curr == '.' {
                value.push(self.curr);
                self.advance();
            }
            let loc = Loc::new(self.file.clone(), start, self.line_index);
            if value.contains(".") {
                return Token::new(value.as_str(), TokenType::Float, loc);
            }
            return Token::new(value.as_str(), TokenType::Integer, loc);
        }

        fn string(&mut self) -> Token {
            let mut value: String = String::new();
            let start = self.line_index;
            while self.curr != '\"' {
                value.push(self.curr);
                self.advance();
            }
            let loc = Loc::new(self.file.clone(), start, self.line_index);
            self.advance();
            return Token::new(value.as_str(), TokenType::String, loc);
        }

        fn char(&mut self) -> Token {
            let mut value: String = String::new();
            let start = self.line_index;
            while self.curr != '\'' {
                value.push(self.curr);
                self.advance();
            }
            let loc = Loc::new(self.file.clone(), start, self.line_index);
            self.advance();
            return Token::new(value.as_str(), TokenType::Char, loc);
        }

        fn id(&mut self) -> Token {
            let mut value: String = String::new();
            let start = self.line_index;
            while self.curr.is_alphanumeric() || self.curr == '_' {
                value.push(self.curr);
                self.advance();
            }
            let loc = Loc::new(self.file.clone(), start, self.line_index);
            if let Some(typ) = KeyWords.get(&value.as_str()) {
                return Token::new(value.as_str(), typ.clone(), loc);
            }

            return Token::new(value.as_str(), TokenType::Id, loc);
        }

        fn advance_token(&mut self, token: Token) -> Token {
            self.advance();
            return token;
        }

        /*
        *   FIXME(S): multi character values
        */

        fn next_token(&mut self) -> Token {
            while self.curr != '\0' {
                while self.curr.is_whitespace() {
                    self.advance();
                }

                let loc = Loc::new(self.file.clone(), self.line_index, self.line_index + 1);
                match self.curr {
                    '.' => return self.advance_token(Token::new(".", TokenType::Dot, loc)),
                    ',' => return self.advance_token(Token::new(",", TokenType::Comma, loc)),
                    ';' => return self.advance_token(Token::new(";", TokenType::Semi, loc)),
                    ':' => match self.src[self.index + 1] {
                        ':' => {
                            self.advance();
                            return self.advance_token(Token::new("::", TokenType::OpScope, loc));
                        }
                        _ => return self.advance_token(Token::new(":", TokenType::Colon, loc)),
                    },
                    '+' => return self.advance_token(Token::new("+", TokenType::OpPlus, loc)),
                    '-' => match self.src[self.index + 1] {
                        '>' => {
                            self.advance();
                            return self.advance_token(Token::new("->", TokenType::Arrow, loc));
                        }
                        _ => return self.advance_token(Token::new("-", TokenType::OpMinus, loc)),
                    },
                    '*' => return self.advance_token(Token::new("*", TokenType::OpMul, loc)),
                    '/' => match self.src[self.index + 1] {
                        '/' => {
                            while self.curr != '\n' {
                                self.advance();
                            }
                            continue;
                        },
                        '*' => {
                            let mut braces = 0;
                            loop {
                                if self.curr == '/' && self.src[self.index + 1] == '*' {
                                    self.advance();
                                    self.advance();
                                    braces += 1;
                                }
                                if self.curr == '*' && self.src[self.index + 1] == '/' {
                                    self.advance();
                                    self.advance();
                                    braces -= 1;
                                    if braces == 0 {
                                        break;
                                    }
                                }
                                self.advance();
                                if self.curr == '\0' {
                                    break;
                                }
                            }
                            continue;
                        },
                        _ => return self.advance_token(Token::new("/", TokenType::OpDiv, loc)),
                    },
                    '%' => return self.advance_token(Token::new("%", TokenType::OpMod, loc)),
                    '[' => return self.advance_token(Token::new("[", TokenType::BraceLeft, loc)),
                    ']' => return self.advance_token(Token::new("]", TokenType::BraceRight, loc)),
                    '{' => return self.advance_token(Token::new("{", TokenType::CurlyLeft, loc)),
                    '}' => return self.advance_token(Token::new("}", TokenType::CurlyRight, loc)),
                    '(' => return self.advance_token(Token::new("(", TokenType::ParanLeft, loc)),
                    ')' => return self.advance_token(Token::new(")", TokenType::ParanRight, loc)),
                    '#' => return self.advance_token(Token::new("#", TokenType::Hash, loc)),
                    '$' => return self.advance_token(Token::new("$", TokenType::Dollar, loc)),
                    '<' => match self.src[self.index + 1] {
                        '=' => {
                            self.advance();
                            return self.advance_token(Token::new("<=", TokenType::OpLeEq, loc));
                        }
                        _ => return self.advance_token(Token::new("<", TokenType::AngleLeft, loc)),
                    },
                    '>' => match self.src[self.index + 1] {
                        '=' => {
                            self.advance();
                            return self.advance_token(Token::new(">=", TokenType::OpGrEq, loc));
                        }
                        _ => {
                            return self.advance_token(Token::new(">", TokenType::AngleRight, loc))
                        }
                    },
                    '&' => match self.src[self.index + 1] {
                        '&' => {
                            self.advance();
                            return self.advance_token(Token::new("&&", TokenType::OpAnd, loc));
                        }
                        _ => return self.advance_token(Token::new("&", TokenType::Ampercent, loc)),
                    },
                    '|' => match self.src[self.index + 1] {
                        '|' => {
                            self.advance();
                            return self.advance_token(Token::new("||", TokenType::OpOr, loc));
                        }
                        _ => return self.advance_token(Token::new("|", TokenType::Pipe, loc)),
                    },
                    '!' => match self.src[self.index + 1] {
                        '=' => {
                            self.advance();
                            return self.advance_token(Token::new("!=", TokenType::OpNeEq, loc));
                        }
                        _ => return self.advance_token(Token::new("!", TokenType::Bang, loc)),
                    },
                    '=' => match self.src[self.index + 1] {
                        '=' => {
                            self.advance();
                            return self.advance_token(Token::new("==", TokenType::OpEqEq, loc));
                        }
                        _ => return self.advance_token(Token::new("=", TokenType::Equals, loc)),
                    },

                    '\"' => {
                        self.advance();
                        return self.string();
                    }
                    '\'' => {
                        self.advance();
                        return self.char();
                    }
                    '\0' => return Token::new("", TokenType::Eof, loc),
                    _ => {
                        if self.curr.is_digit(10) {
                            return self.number();
                        }

                        if self.curr.is_alphabetic() {
                            return self.id();
                        }

                        panic!("unknown token found `{}`", self.curr);
                    }
                }
            }
            return Token::new("", TokenType::Eof, Loc::new(PathBuf::new(), self.line_index, self.line_index + 1));
        }
    }
}
